4.x: Allow BatchStatements to be LWT

Bouncheck opened the pull request:
Previously all DefaultBatchStatements would always return `false` when `isLWT()` was called. This would cause the driver to route the batch based on the first non null routing information found in a batch but using regular rules rather than rules for LWT queries, even if a LWT query was inside the batch.

Now LWT routing will be used for DefaultBatchStatements if somewhere along the way an LWT query was added to the batch. This can also be controlled explicitly regardless of batch contents with
`BatchStatementBuilder#setIsLWT(boolean)`.

Fixes: https://github.com/scylladb/java-driver/issues/488

Bouncheck committed:
Allow BatchStatements to be LWT

Previously all DefaultBatchStatements would always return `false` when `isLWT()`
was called. This would cause the driver to route the batch based on the
first non null routing information found in a batch but using regular rules
rather than rules for LWT queries, even if a LWT query was inside the batch.

Now LWT routing will be used for DefaultBatchStatements if somewhere along
the way an LWT query was added to the batch. This can also be controlled
explicitly regardless of batch contents with
`BatchStatementBuilder#setIsLWT(boolean)`.

Bouncheck commented:
Fixes #488

Bouncheck commented:
When I've checked the current implementation of BatchStatement it turned out that it already infers the routing information from the first query in a batch that has routing information. I've left that part as is and just added LWT support for batches according to suggestions.

Lorak-mmk commented on core/src/test/java/com/datastax/oss/driver/internal/core/cql/DefaultBatchStatementTest.java:
```diff
@@ -97,4 +101,37 @@ public void should_not_issue_log_warn_if_statement_have_no_consistency_level_set
 
     verify(logger.appender, times(0)).doAppend(logger.loggingEventCaptor.capture());
   }
+
+  @Test
+  public void should_inherit_lwt_status() {
+    // SELECT is not allowed in practice but is sufficient for unit testing
+    SimpleStatement simpleStatement =
+        SimpleStatement.builder("SELECT * FROM some_table WHERE a = ?").build();
+    BoundStatement lwtBoundStatement = mock(DefaultBoundStatement.class);
+    when(lwtBoundStatement.isLWT()).thenReturn(true);
+
+    // Without LWT statements added
+    BatchStatementBuilder batchStatementBuilder = new BatchStatementBuilder(BatchType.UNLOGGED);
+    batchStatementBuilder.addStatement(simpleStatement);
+    assertThat(batchStatementBuilder.build().isLWT()).isFalse();
+
+    // Check if implicitly set to true after adding LWT bound statement
+    batchStatementBuilder.addStatement(lwtBoundStatement);
+    assertThat(batchStatementBuilder.build().isLWT()).isTrue();
+
+    // Check if explicit set overrides implicit resolution
+    batchStatementBuilder.setIsLWT(false);
+    assertThat(batchStatementBuilder.build().isLWT()).isFalse();
+    batchStatementBuilder = new BatchStatementBuilder(BatchType.UNLOGGED);
+    batchStatementBuilder.addStatement(simpleStatement);
+    batchStatementBuilder.setIsLWT(true);
+    assertThat(batchStatementBuilder.build().isLWT()).isTrue();
+
+    assertThat(BatchStatement.newInstance(BatchType.UNLOGGED).isLWT()).isFalse();
+    assertThat(BatchStatement.newInstance(BatchType.LOGGED).isLWT()).isFalse();
+    assertThat(BatchStatement.newInstance(BatchType.COUNTER).isLWT()).isFalse();
+    assertThat(BatchStatement.newInstance(BatchType.UNLOGGED, lwtBoundStatement).isLWT()).isTrue();
+    assertThat(BatchStatement.newInstance(BatchType.LOGGED, lwtBoundStatement).isLWT()).isTrue();
+    assertThat(BatchStatement.newInstance(BatchType.COUNTER, lwtBoundStatement).isLWT()).isTrue();
+  }
```
Could you also add integration test that creates and executes batch that contains both LWT and non-LWT statements?

dkropachev replied:
and test that would check that resulted LWT batch get routed properly.

Bouncheck replied:
Will do

Bouncheck replied:
See if `LWTLoadBalancing#should_use_only_one_node_when_lwt_batch_detected` is sufficient. Basically the same as methods above it but for batch statement with non-LWT and LWT query.

dkropachev commented on core/src/main/java/com/datastax/oss/driver/api/core/cql/BatchStatementBuilder.java:
```diff
@@ -155,10 +170,19 @@ public BatchStatement build() {
         node,
-        nowInSeconds);
+        nowInSeconds,
+        resolveIsLWT(statements));
```
I think it would be better to do the following:
1. Make `isLWT null` by default
2. Unless you enforce it via `setIsLWT`it stays null
3. When you call `IsLWT()` if `isLWT` is `null` you resolves it via `resolveIsLWT`

It will help you with all the cases when you set `setIsLWT` and then manipulate with the batch.

Bouncheck replied:
It already is this way in `BatchStatementBuilder`. It's null by default and will stay like that until enforced. The value for instance of `BatchStatement` is only resolved at the final `build()`. There is no getter for `isLWT` in the builder.

Now in `DefaultBatchStatement` it's a normal boolean because it's immutable by design, so i think it makes sense to follow convention there and since it's supposed to be immutable there is no reason to resolve the value on each `isLWT()` call because it has to be the same all the time.

dkropachev replied:
It is immutable, but it has API that creates another copy of it with some changes.
In some cases behavior of this API becomes confusing, example:
```
   BatchStatement batch = new BatchStatementBuilder(BatchType.UNLOGGED).build();
    batch = batch.add(simpleStatement);
    assertThat(batch.isLWT()).isFalse();
    batch = batch.add(lwtBoundStatement);
    assertThat(batch.isLWT()).isTrue();
    batch = batch.clear();
    assertThat(batch.isLWT()).isFalse();
    batch = batch.add(simpleStatement);
    assertThat(batch.isLWT()).isFalse();
```
You would expect it to succeeded, but it fails, because `isLWT` is persisted.

Bouncheck replied:
I left that this way because `clear` does not modify any other field either, but I guess it's true that it's weird in such case.
Would setting it to `false` on clear be acceptable? It would avoid `CqlRequestHandler` potential scan through the statement list.
If not I can make it `Boolean` and resolve on `isLWT()` method call the same way as in builder.

dkropachev replied:
> I left that this way because `clear` does not modify any other field either, but I guess it's true that it's weird in such case. Would setting it to `false` on clear be acceptable? It would avoid `CqlRequestHandler` potential scan through the statement list. If not I can make it `Boolean` and resolve on `isLWT()` method call the same way as in builder.

I see you point, it is valid, but I still think it won't cut it, since creates another problem:
```
   BatchStatement batch = new BatchStatementBuilder(BatchType.UNLOGGED).setIsLWT(true).build();
    batch = batch.add(simpleStatement);
    batch = batch.add(lwtBoundStatement);
    batch = batch.clear();
    assertThat(batch.isLWT()).isTrue();
```
I think if user want to offload LWT resolution they just need to specify it via API, otherwise driver will have to do this job for them, btw we can throw a warning when driver resolves isLWT recomending them to set it specifically.

Bouncheck replied:
I don't think it's a problem. In that case it's reasonably expected. 
Btw, is there a reason to run empty or non-lwt batch as LWT (which means different routing only)?

Bouncheck replied:
Changed to Boolean in a newer version

dkropachev replied:
You need to update the code not to touch isLWT and resolve it only when `isLWT()` is called

Bouncheck commented:
Adjusted integration test to have a batch with non-LWT and LWT query, added a few extra checks to unit test, adjusted docs and BatchStatement interface.

Lorak-mmk commented:
I unfortunately still don't see an integration test that executes a batch with LWT and non-LWT statements. Could you add it in a separate PR?
Right now there is only a test that has both prepared and unprepared LWT statement.
